src/network/ledger.rs

```diff
@@ -812,16 +812,35 @@ impl<N: Network, E: Environment> Ledger<N, E> {
                 return;
             }
 
+            // Filter out any pre-existing block requests for the peer.
+            let mut new_block_heights = Vec::new();
+            if let Some(block_requests) = self.block_requests.read().await.get(&peer_ip) {
+                for block_height in start_block_height..=end_block_height {
+                    if !block_requests.contains_key(&block_height.into()) {
+                        new_block_heights.push(block_height);
+                    }
+                }
+            } else {
+                self.add_failure(peer_ip, format!("Missing block requests for {}", peer_ip)).await;
+            }
+
             // Log each block request to ensure the peer responds with all requested blocks.
-            for block_height in start_block_height..=end_block_height {
-                // If the ledger was reverted, include the expected new block hash for the fork.
-                match ledger_reverted {
-                    true => {
-                        self.add_block_request(peer_ip, block_height, maximum_block_locators.get_block_hash(block_height))
+            if let Some(locked_block_requests) = self.block_requests.write().await.get_mut(&peer_ip) {
+                for block_height in new_block_heights {
+                    // If the ledger was reverted, include the expected new block hash for the fork.
+                    match ledger_reverted {
+                        true => {
+                            self.add_block_request(
+                                peer_ip,
+                                block_height,
+                                maximum_block_locators.get_block_hash(block_height),
+                                locked_block_requests,
```

Comment:
Perhaps it's worth commenting on the lock holding here or above?

Reply:
Isn't the lock's scope in `if let` sufficient? Or do you mean that it's holding for the duration of the loop?

Reply:
The latter, especially as the mutable reference is being passed into another function.